File: UrnModel/onlyAggr.R
k <- 10000
n <- 1000000
originales <- rnorm(k)
cumulos <- originales - min(originales) + 1
cumulos <- round(n * cumulos / sum(cumulos))
diferencia <- n - sum(cumulos)
if (diferencia > 0) {
    for (i in 1:diferencia) {
        p <- sample(1:k, 1)
        cumulos[p] <- cumulos[p] + 1
    }
} else if (diferencia < 0) {
    for (i in 1:-diferencia) {
        p <- sample(1:k, 1)
        if (cumulos[p] > 1) {
            cumulos[p] <- cumulos[p] - 1
        }
    }
}
c <- median(cumulos) # tamanio critico de cumulos
d <- sd(cumulos) / 4 # factor arbitrario para suavizar la curva
rotura <- function(x) {
    return (1 / (1 + exp((c - x) / d)))
}
union <- function(x) {
    return (exp(-x / c))
}
romperse <- function(tam, cuantos) {
    res <- integer()
    for (cumulo in 1:cuantos) {
        if (runif(1) < rotura(tam)) {
            primera <- sample(1:(tam-1), 1)
            segunda <- tam - primera
            res <- c(res, primera, segunda)
        } else {
            res <- c(res, tam)
        }
    }
    return(res)
}
unirse <- function(tam, cuantos) {
    res <- integer()
    for (cumulo in 1:cuantos) {
        if (runif(1) < union(tam)) {
            res <- c(res, -tam) # marcamos con negativo los que quieren unirse
        } else {
            res <- c(res, tam)
        }
    }
    return(res)
}
freq <- as.data.frame(table(cumulos))
names(freq) <- c("tam", "num")
freq$tam <- as.numeric(levels(freq$tam))[freq$tam]
duracion <- 10
digitos <- floor(log(duracion, 10)) + 1
for (paso in 1:duracion) {
    cumulos <- integer()
    for (i in 1:dim(freq)[1]) {
        urna <- freq[i,]
        cumulos <- c(cumulos, unirse(urna$tam, urna$num))
    }
    juntarse <- -cumulos[cumulos < 0] # sacarlos y hacerlos positivos 
    cumulos <- cumulos[cumulos > 0] # los demas se quedan tal cual
    nt <- length(juntarse)
    if (nt > 0) {
        if (nt > 1) {
            np <- floor(nt / 2) # cuantos pares se pueden formar
            juntarse <- sample(juntarse) # ponerles orden aleatorio
            for (i in 1:np) { # sacar los pares formados por el orden
                cumulos <- c(cumulos, juntarse[2*i-1] + juntarse[2*i])
            }
        }
        if (nt %% 2 == 1) { # es una cantidad impar
            cumulos <- c(cumulos, juntarse[length(juntarse)]) # el ultimo
        }
    }
    tl <- paste(paso, "", sep="")
    while (nchar(tl) < digitos) {
        tl <- paste("0", tl, sep="")
    }
    png(paste("p8_ut", tl, ".png", sep=""), width=300, height=300)
    hist(cumulos, breaks=20, main=paste("Paso", paso, "con pura uni\u{00f3}n"), 
         freq=FALSE, ylim=c(0, 0.05),
         xlab="Tama\u{00f1}o", ylab="Frecuencia relativa")
    graphics.off()
    freq <- as.data.frame(table(cumulos))
    names(freq) <- c("tam", "num")
    freq$tam <- as.numeric(levels(freq$tam))[freq$tam]
}

File: UrnModel/onlyAggr.py
# -*- coding: utf-8 -* # para incluir acentos
import numpy as np
from random import randint
 
k = 1000
n = 100000
orig = np.random.normal(size = k)
cumulos = orig - min(orig)
cumulos += 1 # ahora el menor vale uno
cumulos = cumulos / sum(cumulos) # ahora suman a uno
cumulos *= n # ahora suman a n, pero son valores decimales
cumulos = np.round(cumulos).astype(int) # ahora son enteros
diferencia = n - sum(cumulos) # por cuanto le hemos fallado
cambio = 1 if diferencia > 0 else -1
while diferencia != 0:
    p = randint(0, k - 1)
    if cambio > 0 or (cambio < 0 and cumulos[p] > 0): # sin vaciar
        cumulos[p] += cambio
        diferencia -= cambio
assert sum(cumulos) == n
 
c = np.median(cumulos) # tamaño crítico de cúmulos
d = np.std(cumulos) / 4 # factor arbitrario para suavizar la curva
 
from math import exp, floor, log
 
def rotura(x, c, d):
    return 1 / (1 + exp((c - x) / d))
 
def union(x, c):
    return exp(-x / c)
 
from random import random
 
def romperse(tam, cuantos):
    res = []
    for cumulo in range(cuantos):
        if random() < rotura(tam, c, d):
            primera = randint(1, tam - 1)
            segunda = tam - primera
            res += [primera, segunda]
        else:
            res.append(tam) # no rompió
    return res
 
def unirse(tam, cuantos):
    res = []
    for cumulo in range(cuantos):
        if random() < union(tam, c):
            res.append(-tam) # marcamos con negativo los que quieren unirse
        else:
            res.append(tam)
    return res
 
from scipy.stats import itemfreq
from numpy.random import shuffle
import matplotlib.pyplot as plt
 
duracion = 10
digitos = floor(log(duracion, 10)) + 1
 
for paso in range(duracion):
    freq = itemfreq(cumulos)    
    l = freq.shape[0]
    cumulos = []    
    for i in range(l):
        urna = freq[i]
        cumulos += unirse(urna[0], urna[1])
    cumulos = np.asarray(cumulos)
    neg = cumulos < 0
    a = len(cumulos)
    juntarse = -1 * np.extract(neg, cumulos) # sacarlos y hacerlos positivos
    cumulos = np.extract(~neg, cumulos).tolist() # los demás en una lista
    assert a == len(juntarse) + len(cumulos)
    nt = len(juntarse)
    if nt > 0:
        if nt > 1:
            shuffle(juntarse) # orden aleatorio
            for i in range(nt // 2):
                cumulos.append(juntarse[2*i] + juntarse[2*i+1])
        if nt % 2 == 1: # fue una cantidad impar
            cumulos.append(juntarse[-1]) # el ultimo
    plt.hist(cumulos, align = 'right', normed = True)
    plt.xlabel('Tamaño')
    plt.ylabel('Frecuencia relativa')
    plt.ylim(0, 0.05)
    plt.title('Paso {:d} con pura unión'.format(paso + 1))
    plt.savefig('p8p_ut' + format(paso, '0{:d}'.format(digitos)) + '.png')
    plt.close()
